### ISSUE TITLE
innodb_redo_log_capacity should be x but is exactly x

### ISSUE THREAD

**Issue by qeepcologne** [major/MySQLTuner-perl#714]
example output:
innodb_redo_log_capacity should be (=512M) if possible, so InnoDB Redo log Capacity equals 25% of buffer pool size.

config:
# INNODB #
innodb_redo_log_capacity       = 512M

checked in mysql:
mysql> show global variables like "innodb_redo_log_capacity";
+--------------------------+-----------+
| Variable_name            | Value     |
+--------------------------+-----------+
| innodb_redo_log_capacity | 536870912 |
+--------------------------+-----------+

**Comment by qeepcologne** [comment]
btw why is the recommendaton 25% of the buffer pool size?
https://blogs.oracle.com/mysql/post/dynamic-innodb-redo-log-in-mysql-80
says at least 1 hour during peak.
If you have large buffer pool (e.g. in our case 200G, mysql tuner recommend crazy large size (50G), default is only 100M, calculated value based on the written amount of data is ~500M).

**Comment by seebeen** [comment]
Agreed. Lowering the redo log capacity decreased my RAM usage by a lot. And it seems we have less IO overall

**Comment by ravage84** [comment]
We ran into this, too, after upgrading to MySQL 8.0 on Ubuntu 20.04 this week.

> An undersized Redo Log Capacity is problematic and lead to performance issues.
> However, it’s not recommended to oversize the Redo Log either. Redo Log files consume disk space and increases the recovery time in case of a restart (innodb_fast_shutdown=1) or a sudden crash. And it also slows down shutdown when innodb_fast_shutdown=0.

https://blogs.oracle.com/mysql/post/dynamic-innodb-redo-log-in-mysql-80

This was a small DB of 347 MB (384 MB InnoDB buffer pool size), which would place the "desired" redo log capacity at just under 100 MB, which was configured, anyway.

In our case, we need to increase the InnoDB buffer pool size from 128 MB, which placed the  "desired" redo log capacity at just 32 MB at first.

A few thoughts:

1. A redo log of a few hundred MB shouldn't be a disk capacity issue nowadays
2. A too small redo log is likely a bigger issue than a too big one in most cases
3. Using a fixed percentage doesn't scale well. See [the origin of the famous 80 % of RAM recommendation for the InnoDB buffer pool size](https://www.percona.com/blog/80-ram-tune-innodb_buffer_pool_size/). It likely recommends too small redo logs for smaller DBs and too big ones for big DBs. 
4. Making a recommendation for the redo log based on the current InnoDB buffer pool size likely doesn't make much sense when that one needs to be adjusted, anyway.

> my opinion is to remove this remcommendation and replace it based on warning (there is warning in log if the value is too low: `2023-09-17T00:02:09.388631Z 0 [Warning] [MY-013865] [InnoDB] Redo log writer is waiting for a new redo log file. Consider increasing innodb_redo_log_capacity.`

Something like this makes much more sense, IMHO.

**Comment by matthewlenz** [comment]
To confuse the issue even more check out what mysql does to the innodb_redo_log_capacity if you enable innodb_dedicated_server (automatically configures most of the important settings for you)

https://dev.mysql.com/doc/refman/8.0/en/innodb-dedicated-server.html

Spoiler. In our case with a server that has 128GB of memory it sets the innodb_redo_log_capacity to 56.25% of the server memory.

**Comment by simonhampel** [comment]
> Even though it sounds crazy mysqltuner should probably follow Oracle/Mysql's table if innodb_dedicated_server is enabled.

Given that innodb_dedicated_server sets all these values automatically, I would think the better approach for the case where innodb_dedicated_server is enabled, would be to simply advise that the innodb_redo_log_capacity setting be removed if it is manually defined, to allow MySQL to set its own defaults?

**Comment by remc010** [comment]
Had the same bug on the latest version:

`[!!] Ratio InnoDB redo log capacity / InnoDB Buffer pool size (0.334821428571429%): 7.0G / 28.0G should be equal to 25%`

https://github.com/major/MySQLTuner-perl/pull/801

**Comment by bes-internal** [comment]
If I understood everything correctly, then it would be good to use the definition of the required parameter through live measurement as in this post ( https://blogs.oracle.com/mysql/post/dynamic-innodb-redo-log-in-mysql-80 ):
```
SET GLOBAL innodb_monitor_enable = all;

select VARIABLE_VALUE from performance_schema.global_status where VARIABLE_NAME='Innodb_redo_log_current_lsn' into @a;
select sleep(60) into @garb;
select VARIABLE_VALUE from performance_schema.global_status where VARIABLE_NAME='Innodb_redo_log_current_lsn' into @b;
select format_bytes(abs(@a - @b)*60) per_hour;
```
Which will give a certain size in `per_hour`.

Here's how it can be done.
- warn a user that to determine `innodb_redo_log_capacity` it will be run mysqltuner with some param or ask about this interactively
- warn that measurement will require changing the variable `innodb_monitor_enable`  to on and 60 seconds of waiting (there should be an option to customize), which must be done during the period of maximum load on writing usage to the database.
- calculate `innodb_redo_log_capacity` based on the actual value from `per_hour` and reasonable limits as follows:
-- if detected server memory < 2GB  -- Redo Log Capacity = 100M
-- if detected server memory < 8GB  -- Redo Log Capacity rounded to 100M from selected `per_hour` and no more than 1G
-- if detected server memory > 8GB  -- Redo Log Capacity rounded to 1G from selected `per_hour` value and no more than 16 GB

P.S reasonable limits takes from
> Automatic InnoDB Configuration for a Dedicated MySQL Server:
**innodb_redo_log_capacity**
Redo log capacity is calculated according to the number of logical processors available on the server. The formula is (number of available logical processors / 2) GB, with a maximum dynamic default value of 16 GB
(https://dev.mysql.com/doc/refman/8.4/en/innodb-dedicated-server.html)

**Comment by jmrenouard** [comment]
Hi @simonhampel  @ravage84 @Crease29 @bes-internal 

Last version should fix this issue
https://github.com/jmrenouard/MySQLTuner-perl/tree/v2.9.0

**Comment by jmrenouard** [comment]
Hi @Crease29 @EusebiuOprinoiu @bes-internal @matthewlenz @qeepcologne @ravage84 @remc010 @seebeen @simonhampel,

Thanks for reporting this!

Could you please check and test the result with the latest version/patch to verify if it resolves the issue on your end?

MySQLTuner is a volunteer-run project aimed at delivering a high-quality tuning tool. A living and active community is essential to make it happen, and your feedback is greatly appreciated! 😊

**Comment by EusebiuOprinoiu** [comment]
This is fixed for me.

**Comment by bes-internal** [comment]
looks good, thanks!

**Comment by seebeen** [comment]
Fixed
